http-server-netty/src/main/java/io/micronaut/http/server/netty/handler/Http2ServerHandler.java

```diff
@@ -0,0 +1,346 @@
+    private void init(Http2ConnectionHandler connectionHandler) {
+        this.connectionHandler = connectionHandler;
+        streamKey = connectionHandler.connection().newKey();
+    }
+
+    @Override
+    void flush() {
+        // while reading, hold back flushes for efficiency.
+        // Http2ConnectionHandler.readComplete does a flush.
+        if (!reading) {
+            connectionHandler.flush(ctx);
+        }
+    }
+
+    @Override
+    public int onDataRead(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding, boolean endOfStream) throws Http2Exception {
+        MultiplexedStream stream = connectionHandler.connection().stream(streamId).getProperty(streamKey);
+        if (stream == null) {
+            return padding; // data not consumed
+        }
+        return stream.onDataRead(data, endOfStream) + padding;
+    }
+
+    @Override
+    public void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers, int padding, boolean endOfStream) throws Http2Exception {
+        onHeadersRead(ctx, streamId, headers, 0, Http2CodecUtil.DEFAULT_PRIORITY_WEIGHT, false, padding, endOfStream); // defaults from vertx
+    }
+
+    @Override
+    public void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers, int streamDependency, short weight, boolean exclusive, int padding, boolean endOfStream) throws Http2Exception {
+        io.netty.handler.codec.http2.Http2Stream str = connectionHandler.connection().stream(streamId);
+        Http2Stream stream = new Http2Stream(str);
+        Http2Stream existing = str.setProperty(streamKey, stream);
+        if (existing != null) {
+            // ignore trailer and revert the setProperty. should not be hot path
+            str.setProperty(streamKey, existing);
+            return;
+        }
+        stream.onHeadersRead(HttpConversionUtil.toHttpRequest(streamId, headers, true), endOfStream);
+    }
+
+    @Override
+    public void onPriorityRead(ChannelHandlerContext ctx, int streamId, int streamDependency, short weight, boolean exclusive) throws Http2Exception {
+    }
+
+    @Override
+    public void onRstStreamRead(ChannelHandlerContext ctx, int streamId, long errorCode) throws Http2Exception {
+        MultiplexedStream stream = connectionHandler.connection().stream(streamId).getProperty(streamKey);
+        if (stream != null) {
+            Http2Error http2Error = Http2Error.valueOf(errorCode);
+            if (http2Error == null) {
+                http2Error = Http2Error.INTERNAL_ERROR;
+            }
+            stream.onRstStreamRead(HTTP2_ERRORS.get(http2Error));
+        }
+    }
+
+    @Override
+    public void onSettingsAckRead(ChannelHandlerContext ctx) throws Http2Exception {
+
+    }
+
+    @Override
+    public void onSettingsRead(ChannelHandlerContext ctx, Http2Settings settings) throws Http2Exception {
+
+    }
+
+    @Override
+    public void onPingRead(ChannelHandlerContext ctx, long data) throws Http2Exception {
```

Comment:
im not sure if netty already does it internally. but either way this can go into the next pr since this one is already big enough, i will put it on the list.

Reply:
netty does it internally, i added a test.